gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java

```diff
@@ -303,33 +296,84 @@ private static <T> ObjectConstructor<T> newDefaultImplementationConstructor(
      */
 
     if (Collection.class.isAssignableFrom(rawType)) {
-      if (SortedSet.class.isAssignableFrom(rawType)) {
-        return () -> (T) new TreeSet<>();
-      } else if (Set.class.isAssignableFrom(rawType)) {
-        return () -> (T) new LinkedHashSet<>();
-      } else if (Queue.class.isAssignableFrom(rawType)) {
-        return () -> (T) new ArrayDeque<>();
-      } else {
-        return () -> (T) new ArrayList<>();
-      }
+      @SuppressWarnings("unchecked")
+      ObjectConstructor<T> constructor = (ObjectConstructor<T>) newCollectionConstructor(rawType);
+      return constructor;
     }
 
     if (Map.class.isAssignableFrom(rawType)) {
-      if (ConcurrentNavigableMap.class.isAssignableFrom(rawType)) {
-        return () -> (T) new ConcurrentSkipListMap<>();
-      } else if (ConcurrentMap.class.isAssignableFrom(rawType)) {
-        return () -> (T) new ConcurrentHashMap<>();
-      } else if (SortedMap.class.isAssignableFrom(rawType)) {
-        return () -> (T) new TreeMap<>();
-      } else if (type instanceof ParameterizedType
-          && !String.class.isAssignableFrom(
-              TypeToken.get(((ParameterizedType) type).getActualTypeArguments()[0]).getRawType())) {
-        return () -> (T) new LinkedHashMap<>();
-      } else {
-        return () -> (T) new LinkedTreeMap<>();
-      }
+      @SuppressWarnings("unchecked")
+      ObjectConstructor<T> constructor = (ObjectConstructor<T>) newMapConstructor(type, rawType);
+      return constructor;
+    }
+
+    // Unsupported type; try other means of creating constructor
+    return null;
+  }
+
+  private static ObjectConstructor<? extends Collection<? extends Object>> newCollectionConstructor(
+      Class<?> rawType) {
+
+    // First try List implementation
+    if (rawType.isAssignableFrom(ArrayList.class)) {
+      return ArrayList::new;
```

Comment:
Not sure if the constructor references here are a good idea, or if it would be easier to understand if a lambda `() -> new ArrayList<>()` was used to make it clearer which of the constructors is used.

Reply:
I'm fine with either.